src/item.cpp

```diff
@@ -4132,6 +4132,10 @@ bool item::gunmod_compatible( const item& mod, bool alert ) const
     } else if( get_free_mod_locations( mod.type->gunmod->location ) <= 0 ) {
         msg = string_format( _( "Your %1$s doesn't have enough room for another %2$s mod." ), tname().c_str(), _( mod.type->gunmod->location.c_str() ) );
 
+    } else if( ammo_current() == "generic_no_ammo" ) {
+        msg = string_format( _( "You drain the charge from your %s before trying to modify it." ), tname().c_str() );
+        return true;
```

Comment:
That's not the way if-else-if statements work, once one of the if conditions is evaluated as true, latter conditions are skipped. For why it doesn't trigger, my guess would be that the ammo_current() doesn't not return "generic_no_ammo" in the case you are trying to handle.